WordPress/gutenberg
# Experiment: A simple Modules API with no server dependency graph

## What?


This is the first of a series of experiments to understand the requirements of a hypothetical Modules API. More info in the Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/55942

This experiment aims to implement the simplest possible API, with these characteristics:

- An API to register modules using:
  - A module identifier (required)
  - The full URL or the path (required)
  - Whether the module is for the admin, the frontend or both (required)
  - A version (optional)
- An API to enqueue modules.
- No server graph of dependencies.
- All registered modules add an entry in the import map.
- Enqueued modules add a `<script>` tag in the head.

## Why?


To understand the requirements, we can start with the simplest possible API and test it out with the interactive blocks.

## How?


By leveraging a new `Gutenberg_Modules` class that acts as a singleton and a couple of functions to expose the functionality:

- `gutenberg_register_module` to register modules and add entries to the import map.
- `gutenberg_enqueue_module` to enqueue modules.

Then, we print two times using the `wp_head` action:

- The import map.
- The enqueued modules.

## Testing Instructions





- Add some interactive blocks: Navigation, Query or Image block.
- Check that in the frontend, there is an import map script.
- Check that scripts with `type="module"` are loaded by the interactive blocks.
- Check that the `@wordpress/interactivity` module is loaded by the other modules using `import ... from "@wordpress/interactivity"`.
